### PR TITLE
ticket2 changes:

### PR DESCRIPTION
**Pull request by ASelivonchyk**
- added validation of PlayCharacter fields
- validation moved from dao to dto layer
- added exception handler to send custom message
- codestyle fixes

### CODE REVIEW

File: src/main/java/dev/task/dndquest/controller/PlayCharacterController.java [hunk 1]
```diff
@@ -8,30 +8,41 @@
...
+import java.util.stream.Collectors;
 
 @RestController
 @RequestMapping("/api/character")
 @AllArgsConstructor
 public class PlayCharacterController {
-    PlayCharacterService playCharacterService;
-    DtoMapper<PlayCharacter, PlayCharacterResponseDto, PlayCharacterRequestDto> mapper;
+    private PlayCharacterService playCharacterService;
+    private DtoMapper<PlayCharacter, PlayCharacterResponseDto, PlayCharacterRequestDto> mapper;
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/controller/PlayCharacterController.java:24]: what for?

File: src/main/java/dev/task/dndquest/controller/PlayCharacterController.java [hunk 2]
```diff
@@ -8,30 +8,41 @@
...
+    }
+
+    @ExceptionHandler
+    public ResponseEntity<Object> handleMethodArgumentNotValidException
+            (MethodArgumentNotValidException ex) {
+        return ResponseEntity
+                .status(HttpStatus.PRECONDITION_FAILED)
+                .body(ex.getAllErrors().stream()
+                        .map(DefaultMessageSourceResolvable::getDefaultMessage)
+                        .collect(Collectors.joining(System.lineSeparator())));
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/controller/PlayCharacterController.java:46]: Controller Advice

File: src/main/java/dev/task/dndquest/validator/CharacterClassConstraint.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target(ElementType.FIELD)
+@Retention(RetentionPolicy.RUNTIME)
+@Constraint(validatedBy = CharacterClassConstraintValidator.class)
+public @interface CharacterClassConstraint {
+    String message() default "{wrong.character.class}";
+    Class<?>[] groups() default {};
+    Class<? extends Payload>[] payload() default {};
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/validator/CharacterClassConstraint.java:16]: what for?

**Review by andyislegend — CHANGES_REQUESTED** [review]

File: src/main/java/dev/task/dndquest/controller/PlayCharacterController.java [hunk 4]
```diff
@@ -8,30 +8,41 @@
...
     public void createCharacter(@Valid @RequestBody PlayCharacterRequestDto playCharacterDto){
+        PlayCharacter playCharacter = mapper.mapToEntity(playCharacterDto);
         try {
-            PlayCharacter playCharacter = mapper.mapToEntity(playCharacterDto);
-            playCharacterService.save(playCharacter);
+           playCharacterService.save(playCharacter);
         } catch (DBException e) {
             throw new ResponseStatusException(
                     HttpStatus.PRECONDITION_FAILED, e.getMessage(), e);
         }
```
[the remark this answers is not in the document]
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/controller/PlayCharacterController.java]: removed.

File: src/main/java/dev/task/dndquest/controller/PlayCharacterController.java (hunk 2 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/controller/PlayCharacterController.java]: ok

File: src/main/java/dev/task/dndquest/validator/CharacterClassConstraint.java (hunk 3 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/validator/CharacterClassConstraint.java]: custom validator for PlayCharacter field.

File: src/main/java/dev/task/dndquest/exception/GlobalExceptionHandler.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+@ControllerAdvice
+public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
+    private static final String MESSAGE_TAG = "message";
+    private static final String STATUS_TAG = "status";
+
+    @Override
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(
+            MethodArgumentNotValidException ex, HttpHeaders headers,
+            HttpStatusCode status, WebRequest request) {
+        return ResponseEntity
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/exception/GlobalExceptionHandler.java:25]: why return Map and not the real object (DTO?)

File: src/main/java/dev/task/dndquest/exception/GlobalExceptionHandler.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    @ExceptionHandler(DBException.class)
+    public ResponseEntity<Object> handleDBException(
+            DBException ex, WebRequest request) {
+        Map<String, Object> body = new HashMap<>();
+        body.put(MESSAGE_TAG, ex.getMessage());
+        return ResponseEntity
+                .status(HttpStatus.SERVICE_UNAVAILABLE)
+                .body(body);
+    }
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/exception/GlobalExceptionHandler.java:45]: not good. as I said - u need ControllerAdvice

File: src/main/java/dev/task/dndquest/exception/GlobalExceptionHandler.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        return ResponseEntity
+                .status(HttpStatus.PRECONDITION_FAILED)
+                .body(ex.getAllErrors()
+                        .stream()
+                        .map(e -> {
+                            Map<String, Object> map = new HashMap<>();
+                            map.put(STATUS_TAG, status.toString());
+                            map.put(MESSAGE_TAG, e.getDefaultMessage());
+                            return map;})
+                        .collect(Collectors.toList()));
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/exception/GlobalExceptionHandler.java:34]: wtf?

File: src/main/java/dev/task/dndquest/service/impl/PlayCharacterClassServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+@Service
+@AllArgsConstructor
+public class PlayCharacterClassServiceImpl implements PlayCharacterClassService {
+    private final PlayCharacterClassRepository repository;
+
+    @Override
+    public PlayCharacterClass findByName(String name) {
+        return repository.findByName(name)
+               .orElseThrow(() -> new DBException("no such character class in database"));
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/service/impl/PlayCharacterClassServiceImpl.java:18]: DBException is a very poor name. U will have plenty. Use specific naming - CharNotFoundException

File: src/main/java/dev/task/dndquest/validator/CharacterClassConstraintValidator.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+import dev.task.dndquest.service.PlayCharacterClassService;
+import dev.task.dndquest.validator.annotation.CharacterClassConstraint;
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class CharacterClassConstraintValidator
+         implements ConstraintValidator<CharacterClassConstraint, String> {
+    @Autowired
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/validator/CharacterClassConstraintValidator.java:11]: field injection is for poor people. We use constructor wiering

File: src/main/java/dev/task/dndquest/validator/PlayCharacterConstraintValidator.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+                && (value.getDexterity() >= MIN_PLAY_CHAR_VALUE
+                && value.getDexterity() <= MAX_PLAY_CHAR_VALUE)
+                && (value.getConstitution() >= MIN_PLAY_CHAR_VALUE
+                && value.getConstitution() <= MAX_PLAY_CHAR_VALUE)
+                && (value.getIntelligence() >= MIN_PLAY_CHAR_VALUE
+                && value.getIntelligence() <= MAX_PLAY_CHAR_VALUE)
+                && (value.getWisdom() >= MIN_PLAY_CHAR_VALUE
+                && value.getWisdom() <= MAX_PLAY_CHAR_VALUE)
+                && (value.getCharisma() >= MIN_PLAY_CHAR_VALUE
+                && value.getCharisma() <= MAX_PLAY_CHAR_VALUE);
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/validator/PlayCharacterConstraintValidator.java:32]: not maintainable code

**Review by andyislegend — CHANGES_REQUESTED** [review]

File: src/main/java/dev/task/dndquest/exception/GlobalExceptionHandler.java (hunk 5 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/exception/GlobalExceptionHandler.java:22]: created exception dto

File: src/main/java/dev/task/dndquest/exception/GlobalExceptionHandler.java (hunk 7 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/exception/GlobalExceptionHandler.java]: changet map to dto.

File: src/main/java/dev/task/dndquest/service/impl/PlayCharacterClassServiceImpl.java (hunk 8 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/service/impl/PlayCharacterClassServiceImpl.java]: ok

File: src/main/java/dev/task/dndquest/validator/CharacterClassConstraintValidator.java (hunk 9 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/validator/CharacterClassConstraintValidator.java]: ok

File: src/main/java/dev/task/dndquest/validator/PlayCharacterConstraintValidator.java (hunk 10 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/validator/PlayCharacterConstraintValidator.java]: changed to separate field validation

File: src/main/java/dev/task/dndquest/controller/PlayCharacterController.java [hunk 11]
```diff
@@ -23,15 +17,9 @@ public class PlayCharacterController {
...
+    @ResponseStatus(code = HttpStatus.CREATED)
     public void createCharacter(@Valid @RequestBody PlayCharacterRequestDto playCharacterDto){
-        try {
-            PlayCharacter playCharacter = mapper.mapToEntity(playCharacterDto);
-            playCharacterService.save(playCharacter);
-        } catch (DBException e) {
-            throw new ResponseStatusException(
-                    HttpStatus.PRECONDITION_FAILED, e.getMessage(), e);
-        }
+        playCharacterService.save(mapper.mapToEntity(playCharacterDto));
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/controller/PlayCharacterController.java:23]: The controller doesn't have to know about the mapper. Why not do that on a service level?
Try to keep controllers as thin as possible

File: src/main/java/dev/task/dndquest/service/impl/PlayCharacterClassServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+import dev.task.dndquest.exception.CharNotFoundException;
+import dev.task.dndquest.model.entity.PlayCharacterClass;
+import dev.task.dndquest.repository.PlayCharacterClassRepository;
+import dev.task.dndquest.service.PlayCharacterClassService;
+import lombok.AllArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@AllArgsConstructor
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/service/impl/PlayCharacterClassServiceImpl.java:11]: RequireArgsConstructor

**Review by andyislegend — CHANGES_REQUESTED** [review]

File: src/main/java/dev/task/dndquest/service/impl/PlayCharacterClassServiceImpl.java (hunk 12 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/service/impl/PlayCharacterClassServiceImpl.java:11]: Ok

File: src/main/java/dev/task/dndquest/controller/PlayCharacterController.java (hunk 11 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/controller/PlayCharacterController.java]: understood

**Review by andyislegend — APPROVED** [review]

File: src/main/java/dev/task/dndquest/controller/PlayerController.java [hunk 13]
```diff
@@ -1,15 +1,40 @@
...
 public class PlayerController {
+    PlayerService playerService;
+    AuthenticationService authenticationService;
+    JwtProvider jwtProvider;
 
     @GetMapping
-    public String hello(){
+    public String hello(Authentication authentication, HttpServletRequest request){
         return "hello world!";
     }
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/controller/PlayerController.java:26]: wtf?

File: src/main/java/dev/task/dndquest/controller/PlayerController.java [hunk 14]
```diff
@@ -1,15 +1,40 @@
...
+import jakarta.validation.Valid;
+import lombok.AllArgsConstructor;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.core.Authentication;
+import org.springframework.web.bind.annotation.*;
 
 @RestController
 @RequestMapping("/api/player")
+@AllArgsConstructor
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/controller/PlayerController.java:17]: why AllArgs?

File: src/main/java/dev/task/dndquest/controller/PlayerController.java [hunk 15]
```diff
@@ -1,15 +1,40 @@
...
 public class PlayerController {
+    PlayerService playerService;
+    AuthenticationService authenticationService;
+    JwtProvider jwtProvider;
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/controller/PlayerController.java:21]: where are modificators? why jwt provider here?

File: src/main/java/dev/task/dndquest/model/dto/PlayerRequestDto.java [hunk 16]
```diff
@@ -0,0 +1,16 @@
+package dev.task.dndquest.model.dto;
+
+import jakarta.validation.constraints.Pattern;
+import jakarta.validation.constraints.Size;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+@Data
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/model/dto/PlayerRequestDto.java:8]: Data is bad, should not use it

File: src/main/java/dev/task/dndquest/model/entity/Player.java [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+import lombok.*;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+
+import java.util.Collection;
+import java.util.List;
+
+@Entity
+@Data
+@RequiredArgsConstructor
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/model/entity/Player.java:13]: not needed here

File: src/main/java/dev/task/dndquest/security/authentication/AuthenticationServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+    private final PlayerService playerService;
+    private final PasswordEncoder passwordEncoder;
+    private final JwtProvider jwtProvider;
+
+    @Override
+    public AuthenticationResponseDto login(PlayerRequestDto dto)
+            throws PlayerNotFoundException{
+        Player player = playerService.findByLogin(dto.getLogin());
+        if (!passwordEncoder.matches(dto.getPassword(), player.getPassword())) {
+            throw new PlayerNotFoundException();
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/security/authentication/AuthenticationServiceImpl.java:26]: ??

File: src/main/java/dev/task/dndquest/security/jwt/JwtProvider.java [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+@Component
+@RequiredArgsConstructor
+public class JwtProvider {
+    private static final String JWT_REQ_HEADER_NAME = "Authorization";
+    private static final String JWT_REQ_START = "Bearer ";
+    private static final int JWT_REQ_START_POSITION = 7;
+    @Value("${security.jwt.token.secret-key}")
+    private String secretKey;
+    @Value("${security.jwt.token.expire-length}")
+    private long validityTime;
```
Comment (andyislegend) [inline src/main/java/dev/task/dndquest/security/jwt/JwtProvider.java:27]: make it through ConfigurationProperties

File: src/main/resources/application.properties [hunk 20]
```diff
@@ -1,15 +1,17 @@
...
-spring.datasource.driverClassName=org.h2.Driver
-spring.datasource.username=sa
-spring.datasource.password=password
-spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
-spring.h2.console.enabled = true;
+spring.datasource.url=jdbc:postgresql://localhost:5432/data
+spring.datasource.username=postgres
+spring.datasource.password=user
 spring.jpa.show-sql=true
+spring.jpa.hibernate.ddl-auto=update
```
Comment (andyislegend) [inline src/main/resources/application.properties:6]: ?

Comment (andyislegend) [reply src/main/resources/application.properties:6]: we do not do that. u should use liquibase changelogs

**Review by andyislegend — CHANGES_REQUESTED** [review]

File: src/main/java/dev/task/dndquest/controller/PlayerController.java (hunk 13 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/controller/PlayerController.java]: used it for manual testing. Already deleted.

File: src/main/java/dev/task/dndquest/controller/PlayerController.java (hunk 14 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/controller/PlayerController.java]: changed to required

File: src/main/java/dev/task/dndquest/controller/PlayerController.java (hunk 15 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/controller/PlayerController.java]: fixed

File: src/main/java/dev/task/dndquest/model/dto/PlayerRequestDto.java (hunk 16 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/model/dto/PlayerRequestDto.java]: changed for Getter only

File: src/main/java/dev/task/dndquest/security/authentication/AuthenticationServiceImpl.java (hunk 18 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/security/authentication/AuthenticationServiceImpl.java:26]: .findByLogin() can throw exception, thats why it marked as throws. In line 25 it checks if dto has the same pass with player from db, and if no - throw exception. 
Or question in something else?

File: src/main/java/dev/task/dndquest/security/jwt/JwtProvider.java (hunk 19 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/security/jwt/JwtProvider.java]: done

File: src/main/resources/application.properties (hunk 20 above)
Comment (ASelivonchyk) [reply src/main/resources/application.properties]: I'll add liquibase to project

File: src/main/java/dev/task/dndquest/security/authentication/AuthenticationServiceImpl.java (hunk 18 above)
Comment (andyislegend) [reply src/main/java/dev/task/dndquest/security/authentication/AuthenticationServiceImpl.java:26]: qstn is - why PlayerNotFound if password missmatched? It is generally wrong - all unsuccessful attempt to login must result in BadCredentialsException

File: src/main/java/dev/task/dndquest/model/entity/Player.java (hunk 17 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/model/entity/Player.java:13]: I`m using constructor in dto mapper that uses only fields login and password

File: src/main/java/dev/task/dndquest/security/authentication/AuthenticationServiceImpl.java (hunk 18 above)
Comment (ASelivonchyk) [reply src/main/java/dev/task/dndquest/security/authentication/AuthenticationServiceImpl.java:26]: i got u.

**Review by andyislegend — APPROVED** [review]

**Pull request merged**